File: Examples/test-suite/csharp/director_basic_runme.cs
using System;

namespace director_basicNamespace {

public class runme
{
  static void Main() 
  {
    runme r = new runme();
    r.run();
  }

  void run()
  {
    director_basic_MyFoo a = new director_basic_MyFoo();

    if (a.ping() != "director_basic_MyFoo::ping()") {
      throw new Exception ( "a.ping()" );
    }

    if (a.pong() != "Foo::pong();director_basic_MyFoo::ping()") {
      throw new Exception ( "a.pong()" );
    }

    Foo b = new Foo();

    if (b.ping() != "Foo::ping()") {
      throw new Exception ( "b.ping()" );
    }

    if (b.pong() != "Foo::pong();Foo::ping()") {
      throw new Exception ( "b.pong()" );
    }

    A1 a1 = new A1(1, false);
    a1.Dispose();

    {
      MyOverriddenClass my = new MyOverriddenClass();

      my.expectNull = true;
      if (MyClass.call_pmethod(my, null) != null)
        throw new Exception("null pointer marshalling problem");

      Bar myBar = new Bar();
      my.expectNull = false;
      Bar myNewBar = MyClass.call_pmethod(my, myBar);
      if (myNewBar == null)
        throw new Exception("non-null pointer marshalling problem");
      myNewBar.x = 10;

      // Low level implementation check
      my.testSwigDerivedClassHasMethod();

      // These should not call the C# implementations as they are not overridden
      int v;
      v = MyClass.call_nonVirtual(my);
      if (v != 100) throw new Exception("call_nonVirtual broken() " + v);

      v = MyClass.call_nonOverride(my);
      if (v != 101) throw new Exception("call_nonOverride broken() " + v);

      // A mix of overridden and non-overridden
      MyClassEnd myend = new MyClassEnd();
      MyClass mc = myend;

      v = mc.nonVirtual();
      if (v != 202) throw new Exception("mc.nonVirtual() broken " + v);

      v = MyClass.call_nonVirtual(mc);
      if (v != 202) throw new Exception("call_nonVirtual(mc) broken " + v);

      v = MyClass.call_nonVirtual(myend);
      if (v != 202) throw new Exception("call_nonVirtual(myend) broken" + v);

      v = MyClass.call_nonOverride(mc);
      if (v != 101) throw new Exception("call_nonOverride(mc) broken" + v);

      v = MyClass.call_nonOverride(myend);
      if (v != 101) throw new Exception("call_nonOverride(myend) broken" + v);
    }
  }
}

class director_basic_MyFoo : Foo {
  public director_basic_MyFoo() : base() {
  }

  public override string ping() {
    return "director_basic_MyFoo::ping()";
  }
}

class MyOverriddenClass : MyClass {
  public bool expectNull = false;
  public bool nonNullReceived = false;
  public override Bar pmethod(Bar b) {
    if ( expectNull && (b != null) )
      throw new Exception("null not received as expected");
    return b;
  }

  public new bool nonVirtual() {
    throw new Exception("non-virtual overrides virtual method");
  }

  public new virtual bool nonOverride() {
    throw new Exception("non-override overrides virtual method");
  }
}

class MyClassMiddle : MyClass {
  public override int nonVirtual() {
    return 202;
  }
}

class MyClassEnd : MyClassMiddle {
  public new bool nonVirtual() {
    throw new Exception("non-virtual overrides virtual method");
  }

  public new virtual bool nonOverride() {
    throw new Exception("non-override overrides virtual method");
  }
}

}

File: Examples/test-suite/rust/director_basic_runme.rs
mod director_basic;

use director_basic::{Bar, FooDirector, FooDirectorHandle, MyClassDirector, MyClassDirectorHandle};
use std::os::raw::{c_int, c_void};

struct RustFoo;

impl FooDirector for RustFoo {
    fn ping(&mut self) -> String {
        "RustFoo::ping()".to_string()
    }

    fn pong(&mut self) -> String {
        "RustFoo::pong()".to_string()
    }
}

struct RustMyClass;

impl MyClassDirector for RustMyClass {
    fn method(&mut self, _arg0: *mut c_void) {}

    fn vmethod(&mut self, arg0: Bar) -> Bar {
        Bar::new_0(arg0.x() + 30)
    }

    fn pmethod(&mut self, arg0: Option<&mut Bar>) -> Option<Bar> {
        arg0.map(|bar| Bar::new_0(bar.x()))
    }

    fn nonVirtual(&mut self) -> c_int {
        202
    }

    fn nonOverride(&mut self) -> c_int {
        303
    }

    fn Connect(&mut self) {}
}

fn main() {
    let foo = director_basic::Foo::new();
    let _foo_handle = FooDirectorHandle::connect(&foo, RustFoo);
    if foo.ping() != "RustFoo::ping()" {
        panic!("expected Rust Foo ping override");
    }
    if foo.pong() != "RustFoo::pong()" {
        panic!("expected Rust Foo pong override");
    }

    let mut my = director_basic::MyClass::new();
    let _my_handle = MyClassDirectorHandle::connect(&my, RustMyClass);
    let mut bar = Bar::new_0(7);
    let returned = director_basic::MyClass::call_pmethod(Some(&mut my), Some(&mut bar)).expect("expected Bar pointer");
    if returned.x() != 7 {
        panic!("expected pointer return from Rust director");
    }
    let copied = my.cmethod(&bar);
    if copied.x() != 37 {
        panic!("expected value return from Rust director");
    }
    if director_basic::MyClass::call_nonVirtual(Some(&mut my)) != 202 {
        panic!("expected Rust override for virtual method");
    }
    if director_basic::MyClass::call_nonOverride(Some(&mut my)) != 303 {
        panic!("expected Rust override for virtual method");
    }
}
